[ML] Transforms: Fix unsupported boolean filter when cloning

qn895 opened the pull request:
## Summary

This PR fixes https://github.com/elastic/kibana/issues/133651. Previously, when cloning a transform containing boolean filters, an error will occur because the field name is not matched properly or field name doesn't exist in the data view/runtime mappings. Even so, it's still a valid config. This PR fixes so that it switches to the JSON editor.

https://user-images.githubusercontent.com/43350163/182228312-8334c166-fc91-4532-a5db-c01ae3c357be.mov

Also, instead of throwing an error, we will now just log it out for troubleshooting purposes.

### Checklist

Delete any items that are not applicable to this PR.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)


### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx) |


### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)


<!--ONMERGE {"backportTargets":["8.4"]} ONMERGE-->

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

walterra commented on x-pack/plugins/transform/public/app/common/pivot_aggs.ts:
```diff
@@ -315,3 +315,15 @@ export function getEsAggFromAggConfig(
   return result;
 }
+
+export const isFilterBooleanAgg = (config: PivotAggsConfig) => {
+  if (
+    config.agg === 'filter' &&
+    isPopulatedObject(config.subAggs) &&
+    isPopulatedObject(config, ['aggConfig']) &&
+    isPopulatedObject(config.aggConfig, ['filterAgg']) &&
+    config.aggConfig.filterAgg === 'bool'
+  )
+    return true;
+  return false;
+};
```
Interesting that the linting allows `ifs` without curly braces, I thought we had a code style recommendation to always use them 🤔 . In this case I guess you could also refactor to just:

```
export const isFilterBooleanAgg = (config: PivotAggsConfig) => (
  config.agg === 'filter' &&
  isPopulatedObject(config.subAggs) &&
  isPopulatedObject(config, ['aggConfig']) &&
  isPopulatedObject(config.aggConfig, ['filterAgg']) &&
  config.aggConfig.filterAgg === 'bool'
);
```

qn895 replied:
Fixed here https://github.com/elastic/kibana/pull/137773/commits/1b095b0067453705d9363084d9c566c58dcff705

darnautov commented:
@qn895 I'm still able to submit an invalid JSON and crash the entire page. We should check the JSON syntax is valid before applying the filter bool query
<img width="1670" alt="image" src="https://user-images.githubusercontent.com/5236598/182368560-0b0887ac-1e99-484f-b020-008f81cf15e1.png">

qn895 commented:
@darnautov @walterra Thanks for reviewing. I updated based on the comments and added the JSON syntax check before enabling the Apply button. I also added additional functional tests during the cloning process that checks the popover does show up when clicking the edit button for each aggregation field.

qn895 commented:
Running [flaky test suite runner](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/985)...

darnautov commented on x-pack/plugins/transform/public/app/sections/create_transform/components/aggregation_list/popover_form.tsx:
```diff
@@ -266,6 +271,18 @@ export const PopoverForm: React.FC<Props> = ({ defaultData, otherAggNames, onCha
+      }
+    }
+  }, [aggConfigDef]);
+
```
There is no need in state here:

```suggestion
 const isJSONValid = useMemo(() => {
  if (isPivotAggsConfigWithExtra<unknown>(aggConfigDef)) {
      try {
        // @ts-ignore Partial type but we still need to check if parse-able or not
        JSON.parse(aggConfigDef.aggConfig?.aggTypeConfig?.filterAggConfig);
        return true;
      } catch (e) {
        return false;
      }
    }
 }, [aggConfigDef])
```

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/137773/commits/4bffcc0de7165fd03a29af129743100f2e5df802

darnautov commented on x-pack/plugins/transform/public/app/sections/create_transform/components/aggregation_list/popover_form.tsx:
```diff
@@ -110,6 +114,7 @@ export const PopoverForm: React.FC<Props> = ({ defaultData, otherAggNames, onCha
   const [field, setField] = useState<string | string[]>(
     isPivotAggsConfigWithUiSupport(defaultData) ? defaultData.field : ''
   );
+  const [isJSONValid, setIsJSONValid] = useState<boolean>(true);
```
```suggestion
```

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/137773/commits/4bffcc0de7165fd03a29af129743100f2e5df802

darnautov commented on x-pack/plugins/transform/public/app/sections/create_transform/components/aggregation_list/popover_form.tsx:
```diff
@@ -266,6 +271,18 @@ export const PopoverForm: React.FC<Props> = ({ defaultData, otherAggNames, onCha
+    if (isPivotAggsConfigWithExtra<unknown>(aggConfigDef)) {
+      try {
+        // @ts-ignore Partial type but we still need to check if parse-able or not
+        JSON.parse(aggConfigDef.aggConfig?.aggTypeConfig?.filterAggConfig);
```
JSON.parse(false) will run successfully. Consider the following instead: 
```suggestion
        return !!JSON.parse(aggConfigDef.aggConfig?.aggTypeConfig?.filterAggConfig);
```

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/137773/commits/4bffcc0de7165fd03a29af129743100f2e5df802

qn895 commented:
Running [flaky test suite runner](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/995)...

darnautov commented on x-pack/plugins/transform/public/app/sections/create_transform/components/step_define/common/filter_agg/components/filter_agg_form.tsx:
```diff
@@ -39,7 +39,11 @@ export function getSupportedFilterAggs(
+  // but we still support it as JSON
+  // eslint-disable-next-line no-console
+  console.error(`The field ${fieldName} does not exist in the index or runtime fields`);
+  return undefined;
```
nit: redundant code 
```suggestion
  return undefined;
```

darnautov commented on x-pack/plugins/transform/public/app/sections/create_transform/components/aggregation_list/popover_form.tsx:
```diff
@@ -265,6 +266,24 @@ export const PopoverForm: React.FC<Props> = ({ defaultData, otherAggNames, onCha
 
+  const isJSONValid = useMemo(() => {
+    if (
+      isFilterBooleanAgg(aggConfigDef) &&
```
when you **_clone_** the transform with boolean filter, this check returns false and the page crashes

darnautov commented:
hey @qn895, I started debugging and realised that the current approach to fixing the bug bypassed the original design I came up with for nested aggs. We need the following to have a robust and maintainable solution: 

1. Implement the `isValid` method for `filter` agg [here](https://githu
b.com/elastic/kibana/blob/4584a8b570402aa07832cf3e5b520e5d2cfa7166/x-pack/plugins/transform/public/app/sections/create_transform/components/step_define/common/filter_agg/config.ts#L179). Basically it should contain the code you have in `isJSONValid `. Check how this method is [implemented](https://github.com/elastic/kibana/blob/4584a8b570402aa07832cf3e5b520e5d2cfa7166/x-pack/plugins/transform/public/app/sections/create_transform/components/step_define/common/filter_agg/config.ts#L145) for the `range` agg. The result of `isValid` is propagated from child to parent to derive the overall validation status. 
2. The error message about invalid JSON should be rendered in the [FilterEditorForm](https://github.com/elastic/kibana/blob/3730dd0779ed8efd74aee88f57422781ec1ac122/x-pack/plugins/transform/public/app/sections/create_transform/components/step_define/common/filter_agg/components/editor_form.tsx#L13) component.

qn895 commented:
@elasticmachine merge upstream

darnautov commented on x-pack/plugins/transform/public/app/common/pivot_aggs.ts:
```diff
@@ -224,6 +224,13 @@ export interface PivotAggsConfigWithExtra<T> extends PivotAggsConfigWithUiBase {
   helperText?: () => string | undefined;
 }
 
+export const isPivotAggsConfigWithExtra = <T>(arg: unknown): arg is PivotAggsConfigWithExtra<T> => {
+  return (
+    isPopulatedObject(arg, ['AggFormComponent', 'aggConfig']) &&
+    isPopulatedObject(arg.aggConfig, ['aggTypeConfig'])
+  );
+};
+
```
unused code

```suggestion

```

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/137773/commits/6cde9458bb3ef80020abbb7b3024bfa929a165bc

darnautov commented on x-pack/plugins/transform/public/app/common/pivot_aggs.ts:
```diff
@@ -315,3 +322,13 @@ export function getEsAggFromAggConfig(
 
   return result;
 }
+
+export const isFilterBooleanAgg = (config: PivotAggsConfig) => {
+  return (
+    config.agg === 'filter' &&
+    isPopulatedObject(config.subAggs) &&
+    isPopulatedObject(config, ['aggConfig']) &&
+    isPopulatedObject(config.aggConfig, ['filterAgg']) &&
+    config.aggConfig.filterAgg === 'bool'
+  );
+};
```
unused code 

```suggestion
```

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/137773/commits/6cde9458bb3ef80020abbb7b3024bfa929a165bc

darnautov commented on x-pack/plugins/transform/public/app/sections/create_transform/components/step_define/common/filter_agg/components/editor_form.tsx:
```diff
@@ -36,6 +38,17 @@ export const FilterEditorForm: FilterAggConfigEditor['aggTypeConfig']['FilterAgg
+            />
+          </EuiCallOut>
+        </>
+      ) : undefined}
```
according to [react docs](https://reactjs.org/docs/conditional-rendering.html#preventing-component-from-rendering), we should return `null` to prevent rendering

```suggestion
      ) : null}
```

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/137773/commits/6cde9458bb3ef80020abbb7b3024bfa929a165bc

darnautov commented on x-pack/plugins/transform/public/app/sections/create_transform/components/step_define/common/filter_agg/config.ts:
```diff
@@ -199,6 +211,18 @@ export function getFilterAggTypeConfig(
           return this.filterAggConfig !== undefined ? JSON.parse(this.filterAggConfig!) : {};
         },
+        isValid() {
+          if (typeof this.filterAggConfig === 'string') {
+            try {
+              return !!JSON.parse(this.filterAggConfig);
+            } catch (e) {
+              // eslint-disable-next-line no-console
+              console.error(`Invalid JSON for aggregation definition.\n${e}`);
+              return false;
+            }
+          }
+          return true;
+        },
```
Would be great to avoid code repetition. I reckon it's worth creating a reusable `jsonValidator`, similar to what we have with the [`numberValidator`](https://github.com/elastic/kibana/blob/6b5101014132678f6aabe936a9a96fdea2f6dc85/x-pack/packages/ml/agg_utils/src/validate_number.ts#L19).

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/137773/commits/6cde9458bb3ef80020abbb7b3024bfa929a165bc

darnautov reviewed (commented):
Functionally LGTM! 🚀  Left some comments to clean up the code.

darnautov reviewed (approved):
Tested and LGTM!

walterra reviewed (approved):
Great you were able to sort this out! Code LGTM.

qn895 commented:
@elasticmachine merge upstream

qn895 commented:
[Flaky test suite runner](https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/1010) successful in 50 runs ✅

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.4|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/138525">](https://github.com/elastic/kibana/pull/138525)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.0 -->
